### PR TITLE
add rio_tiler.types and fix type hint issues

### PR DESCRIPTION
**Pull request by vincentsarago**
This PR does:
- better define types of the ColorMap objects
- fix type issues for numpy masked array
- create `rio_tiler.types` submodule to store custom type definition

### CODE REVIEW

File: rio_tiler/constants.py [hunk 1]
```diff
@@ -2,16 +2,11 @@
...
 
 import morecantile
 from rasterio.crs import CRS
 
-NumType = Union[float, int]
-BBox = Tuple[float, float, float, float]
-ColorTuple = Tuple[int, int, int, int]
-NoData = Union[float, int, str]
-Indexes = Union[Sequence[int], int]
+from .types import BBox, ColorTuple, Indexes, NoData, NumType  # noqa
```
Comment (vincentsarago) [inline rio_tiler/constants.py:9]: I love breaking change but I like to be nice sometimes!

File: rio_tiler/io/base.py [hunk 2]
```diff
@@ -694,12 +695,12 @@ def _reader(asset: str, *args, **kwargs: Any) -> Dict:
 
         data = multi_values(assets, _reader, lon, lat, **kwargs)
 
-        values = [d for _, d in data.items()]
+        values = numpy.array([d for _, d in data.items()])
```
Comment (vincentsarago) [inline rio_tiler/io/base.py:698]: from #411

**Review by kylebarron — APPROVED** [review]
❤️  types

File: rio_tiler/io/base.py (hunk 2 above)
Comment (kylebarron) [reply rio_tiler/io/base.py:698]: oh I forgot about this PR 😄

File: rio_tiler/colormap.py [hunk 3]
```diff
@@ -53,16 +54,14 @@ def _remove_value(cmap: Dict, idx: Sequence[int]) -> None:
         cmap.pop(i, None)
 
 
-def _update_cmap(cmap: Dict, values: Dict) -> None:
+def _update_cmap(cmap: GDALColorMapType, values: GDALColorMapType) -> None:
     """Update a colormap dict."""
     for i, color in values.items():
-        if len(color) == 3:
-            color += [255]
         cmap[i] = color
```
Comment (vincentsarago) [inline rio_tiler/colormap.py:60]: Length is defined by the type 🤷

File: rio_tiler/colormap.py [hunk 4]
```diff
@@ -74,19 +73,17 @@ def make_lut(colormap: Dict) -> numpy.ndarray:
     """
     lut = numpy.zeros(shape=(256, 4), dtype=numpy.uint8)
     for i, color in colormap.items():
-        lut[int(i)] = color
+        lut[int(i)] = numpy.array(color)
```
Comment (vincentsarago) [inline rio_tiler/colormap.py:76]: not necessary but I think it's help understanding that we put the tuple in the numpy array

File: rio_tiler/colormap.py [hunk 5]
```diff
@@ -293,7 +290,7 @@ def get(self, name: str) -> Dict:
             colormap = numpy.load(cmap)
             assert colormap.shape == (256, 4)
             assert colormap.dtype == numpy.uint8
-            return {idx: value.tolist() for idx, value in enumerate(colormap)}
+            return {idx: tuple(value) for idx, value in enumerate(colormap)}  # type: ignore
```
Comment (vincentsarago) [inline rio_tiler/colormap.py:293]: `value` entry in the numpy array could be of any length (which is why we add an `assert` test) but mypy is not happy ;-)

File: rio_tiler/types.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+NumType = Union[float, int]
+
+BBox = Tuple[float, float, float, float]
+NoData = Union[float, int, str]
+Indexes = Union[Sequence[int], int]
+
+DataMaskType = Tuple[numpy.ndarray, numpy.ndarray]
+
+ColorTuple = Tuple[int, int, int, int]
```
Comment (vincentsarago) [inline rio_tiler/types.py:15]:
```suggestion
ColorTuple = Tuple[int, int, int, int]  # (red, green, blue, alpha)
```

**Pull request merged**